File: 97interleavingString/interleavingString.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3): return False
        else:
            dp = [True, *itertools.repeat(False, len(s2))]
            for column in range(1, len(dp)): dp[column] = dp[column - 1] and s3[column - 1] == s2[column - 1]
            for row in range(1, len(s1) + 1):
                dp[0] = dp[0] and s3[row - 1] == s1[row - 1]
                for column in range(1, len(dp)): dp[column] = dp[column] and s3[row + column - 1] == s1[row - 1] or dp[column - 1] and s3[row + column - 1] == s2[column - 1]
            return dp[-1]
    
    

File: 97interleavingString/interleavingString.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.codePointCount(0, s1.length()) + s2.codePointCount(0, s2.length()) != s3.codePointCount(0, s3.length())) return false;
        else
        {
            final boolean[] dp = new boolean[s2.codePointCount(0, s2.length()) + 1];
            dp[0] = true;
            IntStream.range(1, dp.length).forEach(column -> dp[column] = dp[column - 1] && s3.codePointAt(column - 1) == s2.codePointAt(column - 1));
            IntStream.range(1, s1.codePointCount(0, s1.length()) + 1).forEach(row -> {
                dp[0] = dp[0] && s3.codePointAt(row - 1) == s1.codePointAt(row - 1);
                IntStream.range(1, dp.length).forEach(column -> dp[column] = dp[column] && s3.codePointAt(row + column - 1) == s1.codePointAt(row - 1) || dp[column - 1] && s3.codePointAt(row + column - 1) == s2.codePointAt(column - 1));});
            return dp[dp.length - 1];
        }
    }
}
